### ISSUE TITLE
Cannot iterate (must implement...) on @phpdoc class[]

### ISSUE THREAD

**Issue by andreyik** [kalessil/phpinspectionsea#800]
Problem is similar to "Incorrect inspection error with foreach over iterable  #793" but I'm not sure it same

```
namespace B;
// any class. doesn't reproduce with scalars
/**
 * @param \stdClass[] $array
 */
function aaa(array $array)
{
   //  array_reverse got false positive error 
    foreach (array_reverse($array) as $item) {
        $item->a();
    }
}
```
![phpstorm](https://user-images.githubusercontent.com/6902309/36138538-7fd7db06-10e5-11e8-81ce-de5447d42d08.png)

**Comment by rentalhost** [comment]
@kalessil Hmm... Initially I thought that was a PS issue, but `array_reserve()` returns an `array`, so it basically should works anyway.

**Comment by kalessil** [comment]
Weird stuff, I'll check why this might happen. I assume @andreyik and @rentalhost  you are both on the latest plugin versions, am I correct?

**Comment by andreyik** [comment]
plugin version: 2.3.17
phpstorm version: 2017.3.4

Maybe its ide bug, but there are no errors if plugin is disabled.

**Comment by rentalhost** [comment]
I am running at:
PhpStorm 2018.1 EAP
PhpInspections U1.0.5

**Comment by rentalhost** [comment]
Some hours later:

The `OpenapiResolveUtil.resolveType()` just take the `array` type when you use some scalar (actually it is based on PS `.getType().global()` API). Then `string[]` is read as only `array` that is okay for this inspection. While `\AnyClass[]` will returns like `array|\AnyClass`.

The problem here is that the inspection can't understand that `\AnyClass` is actually `\AnyClass[]`, but I think that we can accept it in case of `array` is present (that is the case!).

By other way, it could create some "true-negatives" when the `@param` is set to `array|\AnyClass` (and not `array|\AnyClass[]`). But it violates the param definition by mixing non-array with array types in a single param (but it is not a rule, unfortunatelly). But I think valid this approach.

I am working on a solution right now.

**Comment by rentalhost** [comment]
**Closing:** fixed by #801.

**Comment by kalessil** [comment]
I'd add the milestone for both PR and original ticket.
